### PR TITLE
remove counters for threads, fix negative counters

### PR DESCRIPTION
**Pull request by CheSema**
Really want to commit this https://github.com/ClickHouse/ClickHouse/pull/47104
So I have to understand how it works. While I read it I found some things.

- remove collecting and sending profile counters for threads
  Thanks to that https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/ThreadStatusExt.cpp#L371 
I'm sure that that counters are useless. Furthermore query_log do not use those counters, it takes counters from thread group.

- I have noticed that some time client return stats from thread with thread_id not zero. Filtration had a bug.

- When client send multiline query ie "INSERT; SELECT; INSERT" with --print-profile-events option, user got negative counters. That happens because server calculate diff between counters. But some reset was missed right after select.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by CheSema** [comment]
QueryThreadLog -- is remain intact.

File: src/Common/ThreadStatus.h [hunk 1]
```diff
@@ -57,45 +57,57 @@ using ThreadStatusPtr = ThreadStatus *;
...
+    /// The first thread created this thread group
+    UInt64 master_thread_id = 0;
 
     ProfileEvents::Counters performance_counters{VariableContext::Process};
     MemoryTracker memory_tracker{VariableContext::Process};
 
-    ContextWeakPtr query_context;
-    ContextWeakPtr global_context;
+    /// Access to the members below has to be in critical section with  mutex
+    mutable std::mutex mutex;
```
Comment (KochetovNicolai) [inline src/Common/ThreadStatus.h:73]: Usually, I would prefer quite the opposite logic: make read-only fields public and constant, and make mutable fields private and protected by a mutex. Ideally, it should be a class invariant so that it would be impossible to use mutable fields without the lock.

File: src/Common/ThreadStatus.h [hunk 2]
```diff
@@ -57,45 +57,57 @@ using ThreadStatusPtr = ThreadStatus *;
...
 
     InternalTextLogsQueueWeakPtr logs_queue_ptr;
     InternalProfileEventsQueueWeakPtr profile_queue_ptr;
```
Comment (KochetovNicolai) [inline src/Common/ThreadStatus.h:76]: So maybe add getters/setters for these fields as well? (and move to the private part).
Also, it's interesting that `logs_queue_ptr` and `client_logs_level` are actually logically connected (initialized in the same place, and are needed for the same feature - writing to text log). It's interesting why they were so separated here...

File: src/Common/ThreadStatus.h [hunk 3]
```diff
@@ -57,45 +57,57 @@ using ThreadStatusPtr = ThreadStatus *;
...
-    std::unordered_set<UInt64> thread_ids;
-    std::unordered_set<ThreadStatusPtr> threads;
-
-    /// The first thread created this thread group
-    UInt64 master_thread_id = 0;
 
     LogsLevel client_logs_level = LogsLevel::none;
 
     String query;
     UInt64 normalized_query_hash = 0;
```
Comment (KochetovNicolai) [inline src/Common/ThreadStatus.h:81]: Let's make a getter and setter for this as well?
Also, we don't need a setter for `normalized_query_hash`. We can calculate it into `query` setter.

Another option is to make a separate structure for mutable stuff and make a common getter that returns this struct + lock.

Comment (CheSema) [reply src/Common/ThreadStatus.h:81]: I thought about that. Seemed like a big step from totally public struct to such techniques. But If you are encouraging such approach lets do it!

File: src/Common/ThreadStatus.h [hunk 4]
```diff
@@ -57,45 +57,57 @@ using ThreadStatusPtr = ThreadStatus *;
...
+    ContextWeakPtr global_context;
+
+    /// Set up at creation, no race when reading
+    FatalErrorCallback fatal_error_callback;
+
+    /// Set of all thread ids which has been attached to the group
+    std::unordered_set<UInt64> thread_ids;
+
+    /// Set of active threads
+    std::unordered_set<ThreadStatusPtr> threads;
```
Comment (KochetovNicolai) [inline src/Common/ThreadStatus.h:109]: That's a bit funny that `thread_ids` are for all threads and `threads` are only for active ones.
btw now I can't find any place where `threads` is used. Looks like before it was used only in `ThreadGroupStatus::getProfileEventsCountersAndMemoryForThreads`, which is removed now

File: src/Common/ThreadStatus.cpp [hunk 5]
```diff
@@ -157,24 +168,11 @@ ThreadStatus::~ThreadStatus()
...
-        ThreadGroupStatus::ProfileEventsCountersAndMemory counters
-        {
-            performance_counters.getPartiallyAtomicSnapshot(),
-            memory_tracker.get(),
-            thread_id
-        };
-
-        std::lock_guard guard(thread_group->mutex);
-        thread_group->finished_threads_counters_memory.emplace_back(std::move(counters));
-        thread_group->threads.erase(this);
```
Comment (KochetovNicolai) [inline src/Common/ThreadStatus.cpp:171]: Shouldn't we do unlink here?
(probably it is moved to some other place, just want to double-check)

File: src/Common/ThreadStatus.h [hunk 6]
```diff
@@ -204,17 +221,11 @@ class ThreadStatus : public boost::noncopyable
     {
         DetachedFromQuery = 0,  /// We just created thread or it is a background thread
         AttachedToQuery,        /// Thread executes enqueued query
-        Died,                   /// Thread does not exist
     };
 
-    int getCurrentState() const
```
Comment (KochetovNicolai) [inline src/Common/ThreadStatus.h:210]: Thank you for removing this. There is no reason why `memory_order_relaxed` was used

File: src/Common/ThreadStatus.h [hunk 7]
```diff
@@ -204,17 +221,11 @@ class ThreadStatus : public boost::noncopyable
     {
         DetachedFromQuery = 0,  /// We just created thread or it is a background thread
         AttachedToQuery,        /// Thread executes enqueued query
-        Died,                   /// Thread does not exist
```
Comment (KochetovNicolai) [inline src/Common/ThreadStatus.h:207]: And thank you for removing this as well.

File: src/Common/ThreadStatus.h (hunk 4 above)
Comment (CheSema) [reply src/Common/ThreadStatus.h:109]: Will remove it then!

File: src/Common/ThreadStatus.h [hunk 8]
```diff
@@ -204,17 +221,11 @@ class ThreadStatus : public boost::noncopyable
...
     };
 
-    int getCurrentState() const
-    {
-        return thread_state.load(std::memory_order_relaxed);
```
Comment (KochetovNicolai) [inline src/Common/ThreadStatus.h:212]: Btw I have no idea why `thread_state` is `atomic<int>` but not `atomic<ThreadState>` (and why `ThreadState` is not an enum class). Can we change it as well (or am I missing something?)

File: src/Common/ThreadStatus.cpp [hunk 9]
```diff
@@ -226,22 +224,12 @@ void ThreadStatus::attachInternalProfileEventsQueue(const InternalProfileEventsQ
 {
     profile_queue_ptr = profile_queue;
 
-    if (!thread_group)
-        return;
+    chassert(thread_group);
```
Comment (KochetovNicolai) [inline src/Common/ThreadStatus.cpp:227]: I would rather throw a logical error as well

File: src/Common/ThreadStatus.h [hunk 10]
```diff
@@ -153,27 +161,32 @@ class ThreadStatus : public boost::noncopyable
...
 
-    String query_id;
+    String query_id_from_query_context;
 
     /// A logs queue used by TCPHandler to pass logs to a client
     InternalTextLogsQueueWeakPtr logs_queue_ptr;
 
     InternalProfileEventsQueueWeakPtr profile_queue_ptr;
 
+    struct TimePoint
```
Comment (KochetovNicolai) [inline src/Common/ThreadStatus.h:171]: Nice!

File: src/Common/ThreadStatus.h [hunk 11]
```diff
@@ -243,39 +254,34 @@ class ThreadStatus : public boost::noncopyable
     ///   query.
     void setInternalThread();
 
-    /// Starts new query and create new thread group for it, current thread becomes master thread of the query
-    void initializeQuery();
-
     /// Attaches slave thread to existing thread group
-    void attachQuery(const ThreadGroupStatusPtr & thread_group_, bool check_detached = true);
+    void attachTo(const ThreadGroupStatusPtr & thread_group_, bool check_detached = true);
```
Comment (KochetovNicolai) [inline src/Common/ThreadStatus.h:258]: Maybe `attachToGroup`? Or just `attachToGroup`?
To make it a bit more similar to `detachGroup`

File: src/Common/ThreadStatus.h [hunk 12]
```diff
@@ -153,27 +161,32 @@ class ThreadStatus : public boost::noncopyable
...
+        UInt64 nanoseconds = 0;
+        UInt64 microseconds = 0;
+        time_t seconds = 0;
+    };
+
     bool performance_counters_finalized = false;
-    UInt64 query_start_time_nanoseconds = 0;
-    UInt64 query_start_time_microseconds = 0;
-    time_t query_start_time = 0;
-    size_t queries_started = 0;
```
Comment (KochetovNicolai) [inline src/Common/ThreadStatus.h:167]: :D

File: src/Common/ThreadStatus.h [hunk 13]
```diff
@@ -153,27 +161,32 @@ class ThreadStatus : public boost::noncopyable
     /// Use it only from current thread
     ContextWeakPtr query_context;
 
-    String query_id;
+    String query_id_from_query_context;
```
Comment (KochetovNicolai) [inline src/Common/ThreadStatus.h:164]: It's interesting that `setQueryId` is called only from `MemoryTrackerThreadSwitcher`.
Maybe after we replace it with a thread group switcher, we can check if any `ThreadStatus` is created without a thread group but with query context (which is a bit strange). If not, we can probably move `query_id` (and query context) to the thread group.
(or even remove query_context and global_context from thread status finally)

Comment (KochetovNicolai) [reply src/Common/ThreadStatus.h:164]: Ah, `attachQueryContext` is already removed. lol

File: src/Common/ThreadStatus.cpp (hunk 5 above)
Comment (CheSema) [reply src/Common/ThreadStatus.cpp:171]: Unlink is done in

if (deleter)
deleter();

File: src/Common/ThreadStatus.h [hunk 14]
```diff
@@ -153,27 +161,32 @@ class ThreadStatus : public boost::noncopyable
...
 
     /// A logs queue used by TCPHandler to pass logs to a client
     InternalTextLogsQueueWeakPtr logs_queue_ptr;
 
     InternalProfileEventsQueueWeakPtr profile_queue_ptr;
 
+    struct TimePoint
+    {
+        void setUp();
+        void SetUp(std::chrono::time_point<std::chrono::system_clock> now);
```
Comment (KochetovNicolai) [inline src/Common/ThreadStatus.h:174]: actually, it looks a bit strange to me : do we really need both `setUp` and `SetUp` (the last one in not in our code style)?

File: src/Interpreters/ThreadStatusExt.cpp [hunk 15]
```diff
@@ -178,6 +208,20 @@ ProfileEvents::Counters * ThreadStatus::attachProfileCountersScope(ProfileEvents
     return prev_counters;
 }
 
+void ThreadStatus::TimePoint::SetUp(std::chrono::time_point<std::chrono::system_clock> now)
+{
+    // query_start_time_{microseconds, nanoseconds} are all constructed from the same time point
+    // to ensure that they are all equal up to the precision of a second.
+    nanoseconds = timeInNanoseconds(now);
```
Comment (KochetovNicolai) [inline src/Interpreters/ThreadStatusExt.cpp:215]: Now I think: do we really need 3 separate fields?
Probably, just store everything in nanoseconds and convert on fly.
E.g. like it is done in Stapwatch.

File: src/Client/ClientBase.cpp [hunk 16]
```diff
@@ -195,25 +200,21 @@ static void incrementProfileEventsBlock(Block & dst, const Block & src)
         }
     };
     std::map<Id, UInt64> rows_by_name;
+
     for (size_t src_row = 0; src_row < src.rows(); ++src_row)
     {
+        /// Filter out threads stats, use thead group stats
```
Comment (KochetovNicolai) [inline src/Client/ClientBase.cpp:206]: Let's add a bit better comment here.
About why do we actually filer everything except thread_grop. (Like, metrics per thread are not needed, and now we do not send it, and it's only for compatibility now),

**Review by KochetovNicolai — COMMENTED** [review]
Generally looks very good

File: src/Interpreters/ThreadStatusExt.cpp [hunk 17]
```diff
@@ -40,13 +40,39 @@ namespace ErrorCodes
     extern const int CANNOT_SET_THREAD_PRIORITY;
 }
 
+const std::vector<UInt64> ThreadGroupStatus::getInvolvedThreadIds() const
```
Comment (serxa) [inline src/Interpreters/ThreadStatusExt.cpp:43]: This const looks excessive
```suggestion
std::vector<UInt64> ThreadGroupStatus::getInvolvedThreadIds() const
```

File: src/Common/ThreadStatus.h (hunk 13 above)
Comment (CheSema) [reply src/Common/ThreadStatus.h:164]: Yes. Now such invariant is vivid.
query_context could be removed from thread _group_. Other threads when attaching to the group copy query_context from it.
You might have meant `remove query_context  from thread status`? if so, I will consider this.
As I see, some time we copy some information from group to thread for the safety of signal listener code. That class took thread status at row pointer from an other thread and it has to be thread safe. The most difficult part in this part of code.

File: src/Common/ThreadStatus.h [hunk 18]
```diff
@@ -57,45 +57,64 @@ using ThreadStatusPtr = ThreadStatus *;
...
+        UInt64 normalized_query_hash = 0;
+    };
 
-    /// The first thread created this thread group
-    UInt64 master_thread_id = 0;
+    SharedData getSharedData()
+    {
+        std::lock_guard lock(mutex);
+        return shared_data;
+    }
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Common/ThreadStatus.h:97]: Maybe this is not needed already at all?

[the remark this answers is not in the document]
Comment (CheSema) [reply src/Common/ThreadStatus.h:97]: I will add comment to the code)

File: src/Interpreters/ProcessList.cpp [hunk 19]
```diff
@@ -203,19 +203,17 @@ ProcessList::insert(const String & query_, const IAST * ast, ContextMutablePtr q
         ProcessListForUser & user_process_list = user_process_list_it->second;
 
         /// Actualize thread group info
+        CurrentThread::attachQueryForLog(query_);
```
Comment (KochetovNicolai) [inline src/Interpreters/ProcessList.cpp:206]: So far I can see that it's the only usage.
Can we just do it for the group below? And remove `attachQueryForLog` method from thread group.

Also, previously, if there were no thread group (I don't know if it was possible), we did not throw an exception. Now it (in theory) can be thrown from `CurrentThread::attachQueryForLog`.

**Review by KochetovNicolai — APPROVED** [review]

File: src/Interpreters/ProcessList.cpp (hunk 19 above)
Comment (CheSema) [reply src/Interpreters/ProcessList.cpp:206]: 1. About query.
That query is written only here and is read in 2 palces: logToQueryThreadLog и SignalListener.
It is safer if SignalListener reads only data from ThreadStatus and do not lock any mutex's. So this is the reason why query is copied to the ThreadStatus.

2. About exception.
You are right, that is change in behaviour. The tests are fine with that and I'm sure that it is right. Now we have an invariant ThreadStatus has a query <=> ThreadStatus attached to a group with query.

Comment (CheSema) [reply src/Interpreters/ProcessList.cpp:206]: UPD:
About exceptions.
The tests have revealed the issue with attachInternalTextLogsQueue. It is called for the thread without any thread group. That happens when SignalListener calls attachInternalTextLogsQueue with the queue from failed thread.
Here Exception with LOGICAL_ERROR is wrong. Fixing that.

**Comment by CheSema** [comment]
stateless ubsan -- 02346_full_text_search -- https://github.com/ClickHouse/ClickHouse/issues/47696

File: src/Server/TCPHandler.cpp [hunk 20]
```diff
@@ -501,6 +501,7 @@ void TCPHandler::runImpl()
             /// (i.e. deallocations from the Aggregator with two-level aggregation)
             state.reset();
             query_scope.reset();
+            last_sent_snapshots.clear();
```
Comment (novikd) [inline src/Server/TCPHandler.cpp:504]: `unordered_map::clear` is not required to release memory. Also, this deallocation should be done before `query_scope.reset()`.

Comment (CheSema) [reply src/Server/TCPHandler.cpp:504]: That diff is not about memory. It has to be cleaned otherwise wrong metrics are send to the client. That line is about `fix negative counters`.

> his deallocation should be done before query_scope.reset()

Why? I do not catch.

**Comment by nickitat** [comment]
guys, are you kidding?

Screenshot 2023-05-30 at 13 37 32

File: src/Processors/Transforms/buildPushingToViewsChain.cpp [hunk 21]
```diff
@@ -286,12 +286,7 @@ Chain buildPushingToViewsChain(
         std::unique_ptr<ThreadStatus> view_thread_status_ptr = std::make_unique<ThreadStatus>();
         /// Copy of a ThreadStatus should be internal.
         view_thread_status_ptr->setInternalThread();
-        /// view_thread_status_ptr will be moved later (on and on), so need to capture raw pointer.
-        view_thread_status_ptr->deleter = [thread_status = view_thread_status_ptr.get(), running_group]
```
Comment (novikd) [inline src/Processors/Transforms/buildPushingToViewsChain.cpp:290]: Why is this line removed?

Comment (CheSema) [reply src/Processors/Transforms/buildPushingToViewsChain.cpp:290]: There is only one deleter function for all cases.
```
thread_status->detachQuery();
```
Was no point to set it explicitly. I removed all those assignments.

File: src/Interpreters/ThreadStatusExt.cpp [hunk 22]
```diff
@@ -537,28 +543,20 @@ void CurrentThread::finalizePerformanceCounters()
...
+    current_thread->detachFromGroup();
 }
 
-
 CurrentThread::QueryScope::QueryScope(ContextMutablePtr query_context, std::function<void()> fatal_error_callback)
 {
-    CurrentThread::initializeQuery();
-    CurrentThread::attachQueryContext(query_context);
     if (!query_context->hasQueryContext())
         query_context->makeQueryContext();
```
Comment (CheSema) [inline src/Interpreters/ThreadStatusExt.cpp:556]: That is the reason of performance degradation in that PR.
It happens that the first thread had an incorrect configuration. Query context had been made right after ThreadStatus attaches to the ThreadGroup. As a result no perf events metrics have been gathered from this first thread.

https://github.com/ClickHouse/ClickHouse/pull/52071

Comment (CheSema) [reply src/Interpreters/ThreadStatusExt.cpp:556]: As follow up I propose turn of metrics_perf_events_enabled for measured runs
https://github.com/ClickHouse/ClickHouse/pull/52072
